File: python solutions/problem_12.py
'''
Program finding the first triangle number with over n divisors.
A triangle number is the sum of the natural numbers up to it.

ex: the 7th triangle number would be 1 + 2 + 3 + 4 + 5 + 6 + 7 = 28 '''

from math import sqrt

def finding_triangle_number(n):
    return n*(n+1)/2

def triangle_divisors(n):   #function returning the first triangle number with more than n divisors
    i = 1
    while(1):
        if(divisors(finding_triangle_number(i)) > n - 1):
            return int(finding_triangle_number(i))
        i += 1

def divisors(n):            #function returning the number of divisors of a number
    result = 0
    for i in range(1, round(sqrt(n)) + 1):
        if n%i == 0:
            result += 2
    return result-1
        

if __name__=="__main__":
    print(triangle_divisors(500))

File: haskell solutions/problem_12.hs
--Program finding the first triangle number with more than n divisors 
--Example: the 7th triangle number is = 1 + 2 + 3 + 4 + 5 + 6 + 7 = 28
--So, the nth number is equal to n*(n + 1)/2

number_of_divisors:: Int -> Int     --function returning the number of factors of a number 
number_of_divisors n  = 
    let list = filter (\x -> rem n x == 0) [1..round(sqrt(fromIntegral n))]
    in 2*(length list)

triangle_number:: Int -> Int -> Int --recursive function returning the first triangle number with more than n divisors
triangle_number n divisors = 
    if number_of_divisors (n*(n + 1) `div` 2) >= divisors 
        then (n*(n + 1) `div` 2)
    else triangle_number (n + 1) divisors

compute:: Int -> Int 
compute n = triangle_number 1 n

main:: IO()
main = do 
    let result = compute 500
    print(result)